Fix pdf exporting issues after refractor

ForeverRainbow opened the pull request:
This fixes several issues introduced with the refractor of pdf exporting. 
- Codeboxes and tables will now extend over several pages.
- Page breaks are now properly calculated as part of required page number.
- Wrapping is now accounted for when calculating the number of pages.

I have tested this with ~200, ~2000, and ~4000 page documents. Note that page breaks are appended on the end of each page, which causes an extra page on the end of any document using page breaks. I do not see this as much of an issue but it would be a trivial fix if needed.


Fixes: #1012

txe commented:
Are you to going to fix this?
>It's too much broken, @ForeverRainbow, can you fix it all and check other things?
> -   document with one empty node, options: All Tree, Include Node Name. Result: generate 3 pages
> -   document with 2 empty nodes, options: All Tree, Include Node Name. Result: generate 3 pages, only one name
> -   document with 2 empty nodes, options: All tree, New Node on New Page. Result: only one page

ForeverRainbow commented:
@txe I didn't explicitly mention it but those are fixed.

txe commented:
I'm asking because i checked and it still didn't work

ForeverRainbow commented:
@txe just checked again, it seems node names are broken for empty nodes, thanks for checking it!

ForeverRainbow commented:
The issue was `pango_get_from_treestore_node` clears its input vector, so the names were getting removed right after being added

txe commented:
Can you remove the last empty sheet from pdf?

txe commented:
Thanks, can you fix a bit more?
- line spacing became smaller and it doesn't look good
- right padding in table cell is 0 and it also doesn't look good.
To compare (text from wiki about ssd)

![Screenshot_2020-07-17_13-52-25](https://user-images.githubusercontent.com/177347/87779551-9eb97e80-c835-11ea-95bd-a4719f022661.png)

![Screenshot_2020-07-17_13-50-14](https://user-images.githubusercontent.com/177347/87779681-dde7cf80-c835-11ea-8c17-806fff28b0a6.png)

ForeverRainbow commented:
@txe I have fixed the tables text problem (and simplified its code). I am not getting that issue with the overlapping characters that you seem to be getting but I am having problems where long lines are not wrapping properly which I am looking into. Could you upload that test document so I can see what it looks like on my machine?

ForeverRainbow commented:
Well I've found the cause of the text wrapping issues. The problem is that a `Pango::Layout` is meant to be a full paragraph and lines will not wrap automatically on their own. When links are embedded into the text it breaks up the layout and causes the clipping (since the layout line is drawn from somewhere further acrosss than it expects). I honestly I am not sure how to fix this. Looking at the documentation for pangomm it looks like the only option may be to do it manually using `Pango::Context` and cairomm, but that would be very complex by the looks. That or roll back this and wait for pango to add link support, or use a different library for pdfs I guess

txe commented:
That's bad because I really like links in pdf. Can you skip the wrapping issue and add fixes for table text padding and line spacing?

ForeverRainbow commented:
@txe I think the line spacing is also a symptom of this, since the layout is fragmented it cannot space each line properly. I will push my table fixes now.

txe commented:
Thank you

txe commented:
I checked, the line spacing and the text wrapping issues were before and are not related to refactoring. I'm going to fix it.

giuspen commented:
Great, thanks @txe

txe commented:
Sorry, it takes so long. I found a nice approach to make code simpler, but used a wrong way to wrap text. It will take two days more to finish it.
